app/controllers/auth/sessions_controller.rb

```diff
@@ -136,4 +136,34 @@ def clear_attempt_from_session
     session.delete(:attempt_user_id)
     session.delete(:attempt_user_updated_at)
   end
+
+  def on_authentication_success(user, security_measure)
+    @on_authentication_success_called = true
+
+    clear_attempt_from_session
+
+    user.update_sign_in!(request, new_sign_in: true)
+    remember_me(user)
+    sign_in(user)
+    flash.delete(:notice)
+
+    LoginActivity.create(
+      user: user,
+      success: true,
+      authentication_method: security_measure,
+      ip: request.remote_ip,
+      user_agent: request.user_agent
+    )
+  end
+
+  def on_authentication_failure(user, security_measure, failure_reason)
+    LoginActivity.create(
+      user: user,
+      success: false,
+      authentication_method: security_measure,
+      failure_reason: failure_reason,
+      ip: request.remote_ip,
+      user_agent: request.user_agent
```

Comment:
This only pops up after correct password input, right? If not, I feel like in many cases it could just be noise with no reasonable action to take from the user's perspective.

Reply:
Indeed. I decided to ignore incorrect password inputs since there would be too many. But if someone has your password and is trying to break 2fa, it could be useful to know.